cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,297 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <deque>
+#include <span>
+
+#include <rapidsmpf/streaming/core/channel.hpp>
+#include <rapidsmpf/streaming/core/coro_utils.hpp>
+#include <rapidsmpf/streaming/core/fanout.hpp>
+#include <rapidsmpf/streaming/core/message.hpp>
+#include <rapidsmpf/streaming/core/node.hpp>
+
+#include <coro/coro.hpp>
+
+namespace rapidsmpf::streaming::node {
+namespace {
+
+/**
+ * @brief Try to allocate memory from the memory types that the message content uses.
+ *
+ * @param msg The message to allocate memory for.
+ * @return The memory types to try to allocate from.
+ */
+constexpr std::span<const MemoryType> try_memory_types(Message const& msg) {
+    auto const& cd = msg.content_description();
+    // if the message content uses device memory, try to allocate from device memory
+    // first, else allocate from host memory
+    return cd.content_size(MemoryType::DEVICE) > 0
+               ? MEMORY_TYPES
+               : std::span<const MemoryType>{
+                     MEMORY_TYPES.begin() + static_cast<std::size_t>(MemoryType::HOST),
+                     MEMORY_TYPES.end()
+                 };
+}
+
+/**
+ * @brief Asynchronously send a message to multiple output channels.
+ *
+ * @param msg The message to broadcast. Each channel receives a shallow
+ * copy of the original message.
+ * @param chs_out The set of output channels to which the message is sent.
+ */
+Node send_to_channels(
+    Context* ctx, Message const& msg, std::vector<std::shared_ptr<Channel>>& chs_out
+) {
+    std::vector<coro::task<bool>> tasks;
+    tasks.reserve(chs_out.size());
+    for (auto& ch_out : chs_out) {
+        // do a reservation for each copy, so that it will fallback to host memory if
+        // needed
+        // TODO: change this
+        auto res = ctx->br()->reserve_or_fail(msg.copy_cost(), try_memory_types(msg)[0]);
+        tasks.push_back(ch_out->send(msg.copy(res)));
```

Comment:
We own (or should own) the message, so one of these copies is redundant (the "last" output channel can just take ownership of the input message).

Reply:
Good point. Let me move the msg to the last channel.